src/canary.rs

```diff
@@ -179,88 +180,124 @@ fn round_up(n: u32, k: u32) -> u32 {
 
-    // store current PC and set PC to beginning of subroutine
-    let previous_pc: RegisterValue = core.read_core_reg(PC)?;
-    core.write_core_reg(PC, start)?;
+    /// Execute the subroutine.
+    ///
+    /// ## Assumptions
+    /// - Expects the [`Core`] to be halted and will leave it halted when the function
+    /// returns.
+    /// - `low_addr` and `size` need to be 4-byte-aligned.
+    pub fn execute(core: &mut Core, low_addr: u64, stack_size: u64) -> Result<(), probe_rs::Error> {
+        assert_subroutine!(low_addr, stack_size, self::SUBROUTINE.len() as u64);
 
-    // execute the subroutine and wait for it to finish
-    core.run()?;
-    core.wait_for_core_halted(TIMEOUT)?;
+        // prepare subroutine
+        let previous_pc = super::prepare_subroutine(core, low_addr, stack_size, self::SUBROUTINE)?;
 
-    // overwrite subroutine
-    core.write_8(start.into(), &[CANARY_VALUE; SUBROUTINE_LENGTH])?;
+        // execute subroutine and wait for it to finish
+        core.run()?;
+        core.wait_for_core_halted(TIMEOUT)?;
 
-    // reset PC to where it was before
-    core.write_core_reg(PC, previous_pc)?;
+        // overwrite subroutine
+        // NOTE: SUBROUTINE.len() must be a multiple of 4, therefore we can safely divide by 4 here
+        core.write_32(low_addr, &[CANARY_U32; self::SUBROUTINE.len() / 4])?;
 
-    Ok(())
-}
+        // reset PC to where it was before
+        core.write_core_reg(PC, previous_pc)
+    }
 
-/// The length of the subroutine.
-const SUBROUTINE_LENGTH: usize = 28;
-
-/// Create a subroutine to paint [`CANARY_VALUE`] from `start` till `end`.
-//
-// Roughly corresponds to following assembly:
-//
-// 00000108 <start>:
-//  108:   4803        ldr r0, [pc, #12]   ; (118 <end+0x2>)
-//  10a:   4904        ldr r1, [pc, #16]   ; (11c <end+0x6>)
-//  10c:   4a04        ldr r2, [pc, #16]   ; (120 <end+0xa>)
-//
-// 0000010e <loop>:
-//  10e:   4281        cmp r1, r0
-//  110:   d001        beq.n   116 <end>
-//  112:   c004        stmia   r0!, {r2}
-//  114:   e7fb        b.n 10e <loop>
-//
-// 00000116 <end>:
-//  116:   be00        bkpt    0x0000
-//  118:   20000100    .word   0x20000100  ; start
-//  11c:   20000200    .word   0x20000200  ; end
-//  120:   aaaaaaaa    .word   0xaaaaaaaa  ; pattern
-fn subroutine(start: u32, end: u32) -> [u8; SUBROUTINE_LENGTH] {
-    // convert start and end address to bytes
-    let [s1, s2, s3, s4] = start.to_le_bytes();
-    let [e1, e2, e3, e4] = end.to_le_bytes();
-
-    const CV: u8 = CANARY_VALUE;
-    [
-        0x03, 0x48, // ldr r0, [pc, #12]
-        0x04, 0x49, // ldr r1, [pc, #16]
-        0x04, 0x4a, // ldr r2, [pc, #16]
-        // <loop>
-        0x81, 0x42, // cmp r1, r0
-        0x01, 0xD0, // beq.n   116 <end>
-        0x04, 0xC0, // stmia   r0!, {r2}
-        0xFB, 0xE7, // b.n 10e <loop>
+    const SUBROUTINE: [u8; 12] = [
+        // <paint>
+        0x88, 0x42, // cmp      r0, r1
+        0x01, 0xd8, // bhi.n    200f4 <end>
```

Comment:
If this is a relative jump, the comment should probably indicate that (the comment looks like an absolute address?)

Reply:
The actual comment was `bhi.n	200f4 <paint+0x8>`, which indicates the relative jump. I though that it is slightly easier to read when I change it, but apparently it creates more confusion.

Reply:
It's the 200f4 that confuses me, because that seems like an absolute value.

Reply:
Okay. I will investigate

Reply:
I assume that the problem with the address being absolute would be, that we jump to the wrong location in memory and therefore our subroutine would crash. Is this correct?

I've played around with it and placed the subroutine in different memory locations. It always worked as expected. Is there another way I can test if this is a problem?

Reply:
Oh no. You are right :(

Reply:
I think the instruction is a relative jump, but the disassembly has translated it into an absolute address based on the location? I don't know thumbv7m well enough to disassemble that instruction in my head, but a 16 bit instruction seems too short for an absolute address.

Reply:
Puh. The failure I observed earlier was due to another problem I introduced while testing.

This [online disassembler](http://shell-storm.org/online/Online-Assembler-and-Disassembler/?opcodes=02d1&arch=arm-t&endianness=little&dis_with_raw=True&dis_with_ins=True#disassembly) decodes `01d8` as `bhi #6`, which looks like a relative jump to me. So it really seems that objdump resolves that address. I couldn't find how to instruct it to not do that.

Reply:
I'd err for keeping the comment like it is, since it is what objdump outputs. We maybe could add a note that it is a relative jump after all.

Reply:
But objdump only says that because you told objdump where the data was located in memory when you disassembled it. This is documentation for piece of relocatable code that can be placed anywhere in memory (it goes in at the bottom of the stack region being painted, right?). So personally I would change it to be `bhi #6 <paint+0x08>`.

Reply:
Also true. I will include that in #327.